src/controllers/controllerengine.cpp

```diff
@@ -98,30 +98,48 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
     }
 }
 
-/* -------- ------------------------------------------------------
-Purpose: Resolves a function name to a QScriptValue including
-            OBJECT.Function calls
-Input:   -
-Output:  -
--------- ------------------------------------------------------ */
-QScriptValue ControllerEngine::resolveFunction(const QString& function) const {
-    QHash<QString, QScriptValue>::const_iterator i =
-            m_scriptValueCache.find(function);
-    if (i != m_scriptValueCache.end()) {
-        return i.value();
-    }
+/* ------------------------------------------------------------------
+Purpose: Turn a snippet of JS into a QScriptValue function.
+         Wrapping it in an anonymous function allows any JS that
+         evaluates to a function to be used in MIDI mapping XML files
+         and ensures the function is executed with the correct
+         'this' object.
+Input:   QString snippet of JS that evaluates to a function,
+         int number of arguments that the function takes
+Output:  QScriptValue of JS snippet wrapped in an anonymous function
+------------------------------------------------------------------- */
+QScriptValue ControllerEngine::wrapFunctionCode(const QString& codeSnippet,
+                                                int numberOfArgs) {
+    QScriptValue wrappedFunction;
 
-    QScriptValue object = m_pEngine->globalObject();
-    QStringList parts = function.split(".");
+    QHash<QString, QScriptValue>::const_iterator i =
+            m_scriptWrappedFunctionCache.find(codeSnippet);
 
-    for (int i = 0; i < parts.size(); i++) {
-        object = object.property(parts.at(i));
-        if (!object.isValid()) {
-            break;
+    if (i != m_scriptWrappedFunctionCache.end()) {
+        wrappedFunction = i.value();
+    } else {
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
+        if (!syntaxIsValid(codeSnippet)) {
+            wrappedFunction = m_pEngine->evaluate(
+                "(function () { print('Syntax error in: "
```

Comment:
Yes, but getting the QScriptValue Error returned would require a going down a rabbit hole of refactoring that I spent most of yesterday puzzling about. checkException() returns a boolean and the functions that call it also generally return booleans, depending on what checkException returns. `m_pEngine->clearExceptions()` has to be called before evaluating the next code snippet, otherwise strange things happen. Without clearing the exceptions, after pressing a button that triggered an exception, the same exception would keep appearing until I used 2-3 other, valid controls but the script functions mapped to those controls would execute fine. Clearing the exceptions from checkExceptions() and syntaxIsValid() is the easiest way to do that, but then the functions that call those don't have access to the QScriptValue Error object. Also, internalExecute() then needs to change to handle the Error objects or it will just generically say they're not a function. Wrapping the error messages in the JS functions returned from wrapFunctionCode() is the most elegant way I came up with to deal with this.